Refactor Home Assistant driver to add fan handler

guodong323 opened the pull request:
# Description

This PR introduces the **initial implementation of the Home Assistant fan driver** based on the VOLTTRON Platform Driver Framework.  
The goal is to enable VOLTTRON to read and write Home Assistant `fan.*` entities through a clean driver interface.

## Summary of Changes

### 1. Implemented Home Assistant Fan Driver (Read/Write)
- Added logic for:
  - Reading fan `state` from Home Assistant
  - Mapping HA values (`on` / `off`) to VOLTTRON (`1` / `0`)
  - Writing state changes (`turn_on` / `turn_off`)
- Added detailed logging and improved exception handling.

### 2. Created Device and Registry Configurations
- Added `devices/home_assistant` configuration:
  - `ip_address`, `port`, `access_token`
  - `driver_type = home_assistant`
  - `registry_config = config://home_assistant_fan.json`
- Added registry file defining a single writable point:
  - Entity ID: `fan.test_fan`
  - Volttron Point Name: `fan_state`
  - Type: `int`
  - Writable: `true`

### 3. Fixed Platform Driver Agent Environment Issues
- Reinstalled Platform Driver Agent cleanly.
- Removed corrupted config entries from the config store.
- Reloaded all configuration files in correct order.
- Ensured the driver is detected and started without errors.

### 4. Fixed `IndentationError` in `home_assistant.py`
- Identified corrupted indentation around line 752.
- Restored the correct logic so the Home Assistant driver loads properly.

### 5. Verified Functionality Through Unit Tests
- All major read/write logic is covered.
- Driver initializes and is correctly loaded by the Platform Driver Agent.

---

# Type of Change

- [x] New feature (adds Home Assistant fan driver)
- [x] Bug fix (fixes corruption in `home_assistant.py`)
- [ ] Breaking change
- [ ] Documentation update

---

# How Has This Been Tested?

### ✔ Unit Tests
- Verified:
  - Value ↔︎ state conversion logic  
  - Registry parsing  
  - Driver initialization  
- All tests pass locally.

### ✔ Driver Startup Validation
- Confirmed via:
vctl config list platform.driver
vctl config get platform.driver devices/home_assistant

- Verified successful startup message:
Starting driver: home_assistant

---

# Checklist

- [x] My code follows project style guidelines  
- [x] I have performed a self-review  
- [x] Comments added for complex logic  
- [x] Documentation updated where relevant  
- [x] No new warnings introduced  
- [x] All unit tests pass locally

guodong323 committed:
Refactor Home Assistant driver to add fan handler

TianchenMin commented:
File: services/core/PlatformDriverAgent/platform_driver/interfaces/home_assistant.py

Overall this refactor already matches the new Home Assistant driver design very well. This version is basically our canonical implementation for the new handler-based architecture (exceptions, HomeAssistantServiceCall, domain handlers, _normalize_value, _send_service_call, handler registry, etc.).

From my side there are no blocking changes required. Please just do the following small cleanups so that we can treat this file as the reference implementation for the team:

Run formatting / linting only

Run black / isort (or the project’s standard formatter) on this file so imports and spacing are consistent with the rest of VOLTTRON.

Fix any trivial lint warnings (unused imports, long lines, etc.) reported by the project’s linter.

Tiny docstring nits (optional, not blocking)

In HomeAssistantDomainHandler.build_service_call and _call_service, you can add a short note that all network I/O is centralized in _send_service_call and that domain handlers must remain pure.

That’s it – no structural changes needed. We will align the other team members’ versions of home_assistant.py to this handler architecture (exceptions, _normalize_value, _send_service_call, DOMAIN_HANDLERS, _set_point flow) rather than the other way around.

guodong323 committed:
Refactor Home Assistant driver: normalize handlers, cleanup _set_point, unify cover logic, improve tests

guodong323 commented:
All requested changes have been completed:

- Applied handler-based normalization using _normalize_value for fan, switch, and unified cover logic
- Refactored _set_point to avoid double-casting and use handler path correctly
- Added logging in _call_service for operation descriptions
- Adjusted docstrings where needed to match behavior
- Ran full formatting with black and isort, resolved minor lint issues
- Verified that unit tests pass successfully

This file is now aligned with the agreed-upon handler architecture and can serve as the canonical implementation for the team.
